From: Borislav Petkov <bp@suse.de>
Date: Sun, 17 Dec 2017 16:37:58 +0100
Subject: x86/CPU: Check speculation control CPUID bit
Patch-mainline: submitted on 2018/1/9
References: bsc#1068032

... and enable the corresponding flags.

Carved out from a patch by Tim Chen <tim.c.chen@linux.intel.com> and
improved.

After microcode reload, we need to check CPUID directly as we don't
update the X86_FEATURE flags after a reload.

[jkosina@suse.cz: context]
Signed-off-by: Borislav Petkov <bp@suse.de>
---
 arch/x86/include/asm/spec_ctrl.h     |    1 +
 arch/x86/kernel/cpu/intel.c          |    2 ++
 arch/x86/kernel/cpu/microcode/core.c |    6 +++++-
 arch/x86/kernel/cpu/spec_ctrl.c      |   16 ++++++++++++++++
 4 files changed, 24 insertions(+), 1 deletion(-)

--- a/arch/x86/include/asm/spec_ctrl.h
+++ b/arch/x86/include/asm/spec_ctrl.h
@@ -95,6 +95,7 @@ void x86_enable_ibrs(void);
 void x86_disable_ibrs(void);
 unsigned int x86_ibrs_enabled(void);
 unsigned int x86_ibpb_enabled(void);
+void x86_spec_check(void);
 
 static inline void x86_ibp_barrier(void)
 {
--- a/arch/x86/kernel/cpu/intel.c
+++ b/arch/x86/kernel/cpu/intel.c
@@ -13,6 +13,7 @@
 #include <asm/msr.h>
 #include <asm/bugs.h>
 #include <asm/cpu.h>
+#include <asm/spec_ctrl.h>
 
 #ifdef CONFIG_X86_64
 #include <linux/topology.h>
@@ -479,6 +480,7 @@ static void init_intel(struct cpuinfo_x8
 	if (cpu_has(c, X86_FEATURE_VMX))
 		detect_vmx_virtcap(c);
 
+	x86_spec_check();
 }
 
 #ifdef CONFIG_X86_32
--- a/arch/x86/kernel/cpu/microcode/core.c
+++ b/arch/x86/kernel/cpu/microcode/core.c
@@ -39,6 +39,7 @@
 #include <asm/microcode.h>
 #include <asm/processor.h>
 #include <asm/cmdline.h>
+#include <asm/spec_ctrl.h>
 
 #define MICROCODE_VERSION	"2.01"
 
@@ -417,8 +418,11 @@ static ssize_t reload_store(struct devic
 		if (!ret)
 			ret = tmp_ret;
 	}
-	if (!ret)
+	if (!ret) {
 		perf_check_microcode();
+		x86_spec_check();
+	}
+
 	mutex_unlock(&microcode_mutex);
 	put_online_cpus();
 
--- a/arch/x86/kernel/cpu/spec_ctrl.c
+++ b/arch/x86/kernel/cpu/spec_ctrl.c
@@ -38,3 +38,19 @@ void x86_enable_ibrs(void)
 		native_wrmsrl(MSR_IA32_SPEC_CTRL, FEATURE_ENABLE_IBRS);
 }
 EXPORT_SYMBOL_GPL(x86_enable_ibrs);
+
+/*
+ * Called after upgrading microcode, check CPUID directly.
+ */
+void x86_spec_check(void)
+{
+	if (boot_cpu_data.x86_vendor == X86_VENDOR_INTEL) {
+		if (cpuid_edx(7) & BIT(26)) {
+			ibrs_state = 1;
+			ibpb_state = 1;
+
+			setup_force_cpu_cap(X86_FEATURE_SPEC_CTRL);
+		}
+	}
+}
+EXPORT_SYMBOL_GPL(x86_spec_check);
